### PR TITLE
MGDSTRM-8899: update README and refactor

### PR DESCRIPTION
**Pull request by showuon**
What I did in this PR:
1. update README to include `ManagedKafkaAlterConfigPolicy` and `ManagedKafkaCreateTopicPolicy` descriptions
2. minor refactor

### CODE REVIEW

File: src/main/java/io/bf2/kafka/common/Config.java [hunk 1]
```diff
@@ -134,7 +139,7 @@ public class Config {
      * in the cluster. If this property is not specified, a default of {@link #DEFAULT_MAX_PARTITIONS}
      * will be used in this class.
      */
-    public static final String MAX_PARTITIONS = "max.partitions";
+    public static final String MAX_PARTITIONS = CREATE_TOPIC_POLICY_PREFIX + "max.partitions";
```
Comment (showuon) [inline src/main/java/io/bf2/kafka/common/Config.java:142]: I guess this is a bug, right? @grdryn

File: src/main/java/io/bf2/kafka/common/Config.java [hunk 2]
```diff
@@ -173,4 +178,19 @@ public class Config {
...
+    }
+
+    public static Optional<Short> brokerDefaultReplicationFactor(Map<String, ?> configs) {
+        return getConfig(DEFAULT_REPLICATION_FACTOR, configs)
+                .map(v -> Short.valueOf(v.toString()));
+    }
+
+    public static Optional<Object> getConfig(String name, Map<String, ?> configs){
+        return Optional.ofNullable(configs.get(name));
+    }
```
Comment (showuon) [inline src/main/java/io/bf2/kafka/common/Config.java:195]: Put these methods into common/Config because they are used in create topic policy and alter config policy, and they are config related methods.

File: src/main/java/io/bf2/kafka/common/rule/MinInsyncReplicasRule.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        if (key.equals(MIN_IN_SYNC_REPLICAS_CONFIG)) {
+            Short minInsyncReplicas = Short.parseShort(val);
+            // if not present, cluster default value taken automatically, otherwise
+            // only allow isr if the defined value >= default isr value configured and <= system replication
+            // factor, as there is no meaning setting this value higher than replication factor.
+            if ((brokerMinInSyncReplicas.isPresent()) &&
+                    (minInsyncReplicas < brokerMinInSyncReplicas.get() || minInsyncReplicas > brokerDefaultReplicationFactor.get())) {
+                return Optional.of(String.format("Topic configured with invalid configs: %s=%s, recommended minimum insync replicas are %d",
+                        MIN_IN_SYNC_REPLICAS_CONFIG, val, brokerMinInSyncReplicas.get()));
+            }
```
Comment (showuon) [inline src/main/java/io/bf2/kafka/common/rule/MinInsyncReplicasRule.java:35]: In my previous implementation, I set `min.insync.replicas` as immutable config. But I found Ger's original validation is better. So put it as a rule for both create topic and alter config policy.

File: src/main/java/io/bf2/kafka/common/Config.java (hunk 1 above)
Comment (grdryn) [reply src/main/java/io/bf2/kafka/common/Config.java:142]: I don't think it's necessarily a bug. I think it has always been an unprefixed property name here, and I didn't change that when changing the prefixes.

Maybe it makes sense to change it, but maybe not. I think this config property might be used for other things like metrics or the admin API server, so if we're going to add the prefix here, then we might need to set both versions (and keep both versions in sync over time). I might be wrong about that.

File: src/main/java/io/bf2/kafka/common/rule/ConfigRules.java [hunk 4]
```diff
@@ -5,14 +5,18 @@
 import io.bf2.kafka.common.Config;
 import org.apache.kafka.common.config.AbstractConfig;
 import org.apache.kafka.common.errors.PolicyViolationException;
+import org.apache.kafka.server.policy.CreateTopicPolicy;
```
Comment (grdryn) [inline src/main/java/io/bf2/kafka/common/rule/ConfigRules.java:8]: This import doesn't appear to be used, is it?

File: src/main/java/io/bf2/kafka/common/rule/MinInsyncReplicasRule.java (hunk 3 above)
Comment (grdryn) [reply src/main/java/io/bf2/kafka/common/rule/MinInsyncReplicasRule.java:35]: I don't think I can take credit for that...probably Mike or Ramesh :slightly_smiling_face:

File: README.adoc [hunk 5]
```diff
@@ -20,6 +20,7 @@ Currently included here are:
 
 - `io.bf2.kafka.authorizer.CustomAclAuthorizer` - an https://kafka.apache.org/documentation/#security_authz[Authorizer] implementation
 - `io.bf2.kafka.topic.ManagedKafkaCreateTopicPolicy` - a https://cwiki.apache.org/confluence/display/KAFKA/KIP-108%3A+Create+Topic+Policy[CreateTopicPolicy] implementation
+- `io.bf2.kafka.config.ManagedKafkaAlterConfigPolicy` - a https://cwiki.apache.org/confluence/display/KAFKA/KIP-133%3A+Describe+and+Alter+Configs+Admin+APIs#KIP133:DescribeandAlterConfigsAdminAPIs-AlterConfigs[AlterConfigPolicy] implementation
```
Comment (grdryn) [inline README.adoc:23]:
```suggestion
- `io.bf2.kafka.config.ManagedKafkaAlterConfigPolicy` - an https://cwiki.apache.org/confluence/display/KAFKA/KIP-133%3A+Describe+and+Alter+Configs+Admin+APIs#KIP133:DescribeandAlterConfigsAdminAPIs-AlterConfigs[AlterConfigPolicy] implementation
```

File: src/main/java/io/bf2/kafka/common/rule/MinInsyncReplicasRule.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    private final Optional<Short> brokerMinInSyncReplicas;
+    private final Optional<Short> brokerDefaultReplicationFactor;
+
+    public MinInsyncReplicasRule(Optional<Short> brokerMinInSyncReplicas, Optional<Short> brokerDefaultReplicationFactor) {
+        this.brokerMinInSyncReplicas = brokerMinInSyncReplicas;
+        this.brokerDefaultReplicationFactor = brokerDefaultReplicationFactor;
+    }
+
+    @Override
+    public Optional<String> validate(String key, String val) {
```
Comment (grdryn) [inline src/main/java/io/bf2/kafka/common/rule/MinInsyncReplicasRule.java:25]: Do you think unit tests would be useful for this?

**Review by grdryn — APPROVED** [review]

File: README.adoc [hunk 7]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+      kas.policy.topic-config.enforced=compression.type:producer,segment.jitter.ms:0
+      kas.policy.topic-config.mutable=cleanup.policy,delete.retention.ms,retention.bytes,retention.ms
+      kas.policy.topic-config.range=segment.bytes:52428800:,segment.ms:600000:
+----
+
+Here are the configuration description:
+
+* `kas.policy.create-topic.max.partitions`: Used to specify the upper limit of partitions that should be allowed in the cluster. If this property is not specified, a default of -1 (no limit) is set
+* `kas.policy.create-topic.partition-counter.timeout-seconds`: Used to specify the number of seconds to use as a timeout duration when listing and describing topics method. Default is 10.
+* `kas.policy.create-topic.partition-counter.private-topic-prefix`: Used to specify the topic prefix to match for private/internal topics. Default is no prefix.
```
Comment (SamBarker) [inline README.adoc:129]: In general `Used to specify` -> `Specifies`.

```suggestion
* `kas.policy.create-topic.partition-counter.private-topic-prefix`: Specifies the prefix  used to match against topic name to detect private/internal topics. Defaults to blank - no prefix.
```

File: README.adoc [hunk 8]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+* `kas.policy.create-topic.partition-counter.schedule-interval-seconds`: Used to specify the interval (in seconds) at which to schedule partition counts. Default is 15.
```
Comment (SamBarker) [inline README.adoc:130]:
```suggestion
* `kas.policy.create-topic.partition-counter.schedule-interval-seconds`: Specifies the interval (in seconds) between partition counts. Defaults to 15.
```

File: README.adoc [hunk 9]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+* `kas.policy.create-topic.partition-limit-enforced`: Used to allow disabling of partition limit enforcement. Default is false.
```
Comment (SamBarker) [inline README.adoc:131]:
```suggestion
* `kas.policy.create-topic.partition-limit-enforced`: `true` disables the enforcement of the partition limit. Defaults to `false` (limit enforced).
```

File: README.adoc [hunk 10]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+* `kas.policy.topic-config.enforced`: Used to specify the configs with permitted value. It's a list in the form [keyA]:[valueA],[keyB]:[valueB]. Default value is: "flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"
```
Comment (SamBarker) [inline README.adoc:132]:
```suggestion
* `kas.policy.topic-config.enforced`: Specifies a mapping between a topic policy key and a value. The property is expressed as a comma separated list of `key:value` pairs where the key and value are separated by  a `:`. Hence the it takes the form `keyA:value1,keyB:value2`. Default value is: "flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"
```

I found the use of `[]` around the key and value in the format description confusing as its trying to describe the parsing rules including extra tokens felt misleading. I'm easily persuaded that everyone knows they imply optional and thus should be included...

File: README.adoc [hunk 11]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+* `kas.policy.topic-config.mutable`: Used to specify the configs that allow to be updated. It's a comma separated list. We'll merge everything in kas.policy.topic-config.enforced and kas.policy.topic-config.range config. Default value is: "message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"
```
Comment (SamBarker) [inline README.adoc:133]:
```suggestion
* `kas.policy.topic-config.mutable`: Specifies which properties of topic are allowed to be configured. It's a comma separated list. The list will be merged with `kas.policy.topic-config.enforced` and `kas.policy.topic-config.range config`. Defaults to: "message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"
```

How does the merge work? IIRC an entry appearing in `kas.policy.topic-config.enforced` is enough but the idea of merging could be read as implying that you need to list all config keys here (the super set of config keys) and then the enforcement detail comes from the additional keys.

File: README.adoc [hunk 12]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+* `kas.policy.topic-config.range`: Used to specify the configs that allow values within a range with the format "configA:minA:maxA,configB:minB:maxB,...". Either min or max can be empty, which means we only set lower/upper bound. Default value is: "max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:"
```
Comment (SamBarker) [inline README.adoc:134]:
```suggestion
* `kas.policy.topic-config.range`: Specifies a mapping from config key to an inclusive/closed range off values. Entries are formatted as a triplet using `:` as a separator, `key:min:max` where either min or max are optional implying a single bound as appropriate. Multiple triplets are comma separated. For example  `configA:minA:maxA,configB:minB:maxB,...`. Defaults to: "max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:"
```

File: README.adoc [hunk 13]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
 
+== ManagedKafkaCreateTopicPolicy
+
+The objective of this policy is to:
+
+* validate the topic replication factor (default allows 3 only)
+* validate number of partitions in the cluster
+* validate the create topic configs
+
+To config the create topic policy, you should add config in Strimzi's Kafka CR.
```
Comment (SamBarker) [inline README.adoc:103]: I think I would pull this out as its own section. `Configuring the policies`

File: README.adoc [hunk 14]
```diff
@@ -91,6 +92,57 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+* `kas.policy.topic-config.mutable`: Used to specify the configs that allow to be updated. It's a comma separated list. We'll merge everything in kas.policy.topic-config.enforced and kas.policy.topic-config.range config. Default value is: "message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"
+* `kas.policy.topic-config.range`: Used to specify the configs that allow values within a range with the format "configA:minA:maxA,configB:minB:maxB,...". Either min or max can be empty, which means we only set lower/upper bound. Default value is: "max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:"
+
+== ManagedKafkaAlterConfigPolicy
+
+The objective of this policy is to validate the alter topic configs
+
+The configuration and the description is the same as ManagedKafkaCreateTopicPolicy, except ManagedKafkaAlterConfigPolicy only accepts topic-config configs, i.e.
```
Comment (SamBarker) [inline README.adoc:140]:
```suggestion
ManagedKafkaAlterConfigPolicy shares its configuration with the ManagedKafkaCreateTopicPolicy, that is to say there is one shared policy definition. The alter topic policy only enforces the configuration defined in the following
```

I'm not 100% convinced about this form of words.

File: src/main/java/io/bf2/kafka/common/Config.java (hunk 1 above)
Comment (SamBarker) [reply src/main/java/io/bf2/kafka/common/Config.java:142]: If we are leaving it without a prefix we need to update the readme to match.

File: src/main/java/io/bf2/kafka/common/PartitionCounter.java [hunk 15]
```diff
@@ -157,7 +157,7 @@ public int getMaxPartitions() {
 
     /**
      * @return the value of the {@link Config#PRIVATE_TOPIC_PREFIX} key in the broker configs, or a default of
-     *         {@link Config#DEFAULT_PRIVATE_TOPIC_PREFIX} if not set.
+     *         {@link Config#DEFAULT_NO_PRIVATE_TOPIC_PREFIX} if not set.
```
Comment (SamBarker) [inline src/main/java/io/bf2/kafka/common/PartitionCounter.java:160]: Is this rename intended? I'm not convinced that the change in name is particularly helpful.

**Review by SamBarker — COMMENTED** [review]
I think using a table to describe the config properties would really help. 

Given this PR is mostly about the readme, I've focused on the language there but none of my suggestions should be taken as instructions they are just suggestions.

File: src/main/java/io/bf2/kafka/common/Config.java (hunk 1 above)
Comment (k-wall) [reply src/main/java/io/bf2/kafka/common/Config.java:142]: This property is also used by MetricsManager in kas-fleetshard, so keeping it without the plugin specific prefix is required.,

Comment (showuon) [reply src/main/java/io/bf2/kafka/common/Config.java:142]: I see. change reverted. Thanks.

**Comment by k-wall** [comment]
@showuon could you rebase this PR and review @SamBarker 's suggestion so we can re-review?  thank you.

File: README.adoc (hunk 10 above)
Comment (showuon) [reply README.adoc]: Good suggestion.

File: README.adoc [hunk 16]
```diff
@@ -91,6 +92,66 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+|kas.policy.topic-config.range |"max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:" |Specifies a mapping from config key to an inclusive/closed range off values. Entries are formatted as a triplet using `:` as a separator, `key:min:max` where either min or max are optional implying a single bound as appropriate. Multiple triplets are comma separated. For example  `configA:minA:maxA,configB:minB:maxB,...`.
+|===
+
+[1]"flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"
+[2]"message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"
```
Comment (showuon) [inline README.adoc:120]: Put these 2 value out of the table, to avoid occupying too many table width size.

File: README.adoc [hunk 17]
```diff
@@ -91,6 +92,66 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+
+[cols="2,1,5"]
+|===
+|property |default value |description
+
+|kas.policy.create-topic.max.partitions |-1 |Specifies the upper limit of partitions that should be allowed in the cluster. If this property is not specified, the default of `-1` implies no limit.
+|kas.policy.create-topic.partition-counter.timeout-seconds |10 |Specifies the duration in seconds to use as a timeout when listing and describing topics. Defaults to 10.
+|kas.policy.create-topic.partition-counter.private-topic-prefix |"" |Specifies the prefix used to match against topic name to detect private/internal topics. Defaults to blank - no prefix.
+|kas.policy.create-topic.partition-counter.schedule-interval-seconds |15 |Specifies the interval (in seconds) between partition counts. Defaults to 15.
+|kas.policy.create-topic.partition-limit-enforced |false |`true` disables the enforcement of the partition limit. Defaults to `false` (limit enforced).
```
Comment (SamBarker) [inline README.adoc:113]: Is this right, the description is the opposite mapping to what I expect from the property name. 

I expect `true` -> `Enforced`, `false` -> `Disabled` but your description is the opposite.

File: README.adoc [hunk 18]
```diff
@@ -91,6 +92,66 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+
+* validate the topic replication factor (default allows 3 only)
+* validate number of partitions in the cluster
+* validate the create topic configs
+
+Here are the configuration description:
+
+[cols="2,1,5"]
+|===
+|property |default value |description
```
Comment (SamBarker) [inline README.adoc:107]: Personally I'd have 
```suggestion
|property |description |default value
```

File: README.adoc (hunk 16 above)
Comment (SamBarker) [reply README.adoc:120]: Re-ordering the columns so the description is before the default might make that less of an issue.

File: README.adoc [hunk 19]
```diff
@@ -91,6 +92,66 @@ When a rule matches, the Authorizer will return that rule’s evaluation as a de
...
+|kas.policy.topic-config.range |"max.message.bytes::1048588,segment.bytes:52428800:,segment.ms:600000:" |Specifies a mapping from config key to an inclusive/closed range off values. Entries are formatted as a triplet using `:` as a separator, `key:min:max` where either min or max are optional implying a single bound as appropriate. Multiple triplets are comma separated. For example  `configA:minA:maxA,configB:minB:maxB,...`.
+|===
+
+[1]"flush.ms:9223372036854775807,index.interval.bytes:4096,compression.type:producer,flush.messages:9223372036854775807,segment.jitter.ms:0,min.cleanable.dirty.ratio:0.5,file.delete.delay.ms:60000,segment.index.bytes:10485760,preallocate:false,unclean.leader.election.enable:false"
+[2]"message.timestamp.difference.max.ms,message.timestamp.type,retention.bytes,min.compaction.lag.ms,cleanup.policy,max.compaction.lag.ms,delete.retention.ms,message.downconversion.enable"
+
+=== Configuring the policies
+To config the create topic policy, you should add config in Strimzi's Kafka CR.
```
Comment (SamBarker) [inline README.adoc:123]:
```suggestion
To configure the create topic policy, you should add config in the Kafka CR.
```

File: README.adoc (hunk 18 above)
Comment (SamBarker) [reply README.adoc:107]:
```suggestion
|property |description |default value  | default interpretation 
```
re Reading the descriptions. They all describe what the default is, duplicating the column, and generally what that means. for example from `kas.policy.create-topic.max.partitions`: `If this property is not specified, the default of ``-1`` implies no limit.`. It might be clearer to express that using 4 columns. 

```
|kas.policy.create-topic.max.partitions | Specifies the upper limit of partitions that should be allowed in the cluster.| -1 | If this property is not specified, the default of `-1` implies no limit.
```

**Review by SamBarker — APPROVED** [review]
The table is a big improvement. 

4 columns might work better than 3 but I'd be happy enough to stick with 3 if that is seen as better.

File: README.adoc (hunk 17 above)
Comment (showuon) [reply README.adoc:113]: You're right! Updated.

**Review by SamBarker — APPROVED** [review]
LGTM

**Comment by showuon** [comment]
PR Merged since it's just README update. If there are other comments, I can address in a separate PR. Thanks.
